### PR TITLE
Underlay props from useOverlay to FF text highlight bug

### PR DESCRIPTION
**Pull request by snowystinger**
Closes 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

**Comment by devongovett** [comment]
What if we combined this and #1470? Instead of passing an extra ref or making a new hook, we could return a new set of props from `useOverlay` for the underlay. Then this could handle if you clicked on the underlay and prevent default.

```js
let {overlayProps, underlayProps} = useOverlay(...);

// in underlayProps
let onPointerDown = e => {
  if (e.target === e.currentTarget) {
    e.preventDefault();
  }
}
```

**Review by LFDanLu — APPROVED** [review]
LGTM, tested in FF, fixes the issue in the original PR mentioned here: https://github.com/adobe/react-spectrum/pull/1468#issuecomment-822853961. Other browsers seem fine as well

File: packages/@react-aria/overlays/test/useOverlay.test.js [hunk 1]
```diff
@@ -11,57 +11,101 @@
...
 import React, {useRef} from 'react';
 import {useOverlay} from '../';
 
 function Example(props) {
   let ref = useRef();
-  let {overlayProps} = useOverlay(props, ref);
-  return <div ref={ref} {...overlayProps} data-testid={props['data-testid'] || 'test'}>{props.children}</div>;
+  let {overlayProps, underlayProps} = useOverlay(props, ref);
+  return (
+    <div {...mergeProps(underlayProps, props.underlayProps || {})}>
```
Comment (ktabors) [inline packages/@react-aria/overlays/test/useOverlay.test.js:23]: Why isn't `useOverlay` handling the merging of `props.underlayProps` into the returned `underlayProps`.

File: packages/@react-aria/overlays/test/useOverlay.test.js [hunk 2]
```diff
@@ -11,57 +11,101 @@
...
-    let res = render(
-      <Example isOpen>
-        <input autoFocus data-testid="input" />
-      </Example>
-    );
-
-    let input = res.getByTestId('input');
-    expect(document.activeElement).toBe(input);
-  });
+  describe.each`
```
Comment (ktabors) [inline packages/@react-aria/overlays/test/useOverlay.test.js:32]: Shouldn't we do onTouch and onClick too?

File: packages/@react-spectrum/overlays/src/Modal.tsx [hunk 3]
```diff
@@ -58,10 +58,9 @@ let typeMap = {
 
 let ModalWrapper = forwardRef(function (props: ModalWrapperProps, ref: RefObject<HTMLDivElement>) {
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  let {children, isOpen, type, isDismissable, isKeyboardDismissDisabled, ...otherProps} = props;
+  let {children, isOpen, type, overlayProps, ...otherProps} = props;
```
Comment (ktabors) [inline packages/@react-spectrum/overlays/src/Modal.tsx:61]: This looks like an API change, but I think it is low enough that it doesn't count.

Comment (snowystinger) [reply packages/@react-spectrum/overlays/src/Modal.tsx:61]: it's internal :)

File: packages/@react-aria/overlays/test/useOverlay.test.js (hunk 1 above)
Comment (snowystinger) [reply packages/@react-aria/overlays/test/useOverlay.test.js:23]: there's no such thing as props.underlayProps, it's only for this test

File: packages/@react-aria/overlays/test/useOverlay.test.js (hunk 2 above)
Comment (snowystinger) [reply packages/@react-aria/overlays/test/useOverlay.test.js:32]: I don't understand what you're asking, this tests mouse/pointer/touch

File: packages/@react-aria/overlays/src/useOverlay.ts [hunk 4]
```diff
@@ -106,10 +115,20 @@ export function useOverlay(props: OverlayProps, ref: RefObject<HTMLElement>): Ov
...
+    }
+  };
+
   return {
     overlayProps: {
       onKeyDown,
       ...focusWithinProps
+    },
+    underlayProps: {
+      onPointerDown: onPointerDownUnderlay
```
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-aria/overlays/src/useOverlay.ts:131]: this is the FF bug, on pointer down, which is why when we stopped using pointer events in that other PR, it fixed the bug. unfortunately doing that brought on some different bugs, so we had to re-add pointer events, and now we just handle that pointer event here

**Review by ktabors — APPROVED** [review]

File: packages/@react-aria/overlays/src/useOverlay.ts [hunk 5]
```diff
@@ -106,10 +115,20 @@ export function useOverlay(props: OverlayProps, ref: RefObject<HTMLElement>): Ov
     }
   });
 
+  let onPointerDownUnderlay = e => {
+    // fixes a firefox issue that starts text selection
```
Comment (devongovett) [inline packages/@react-aria/overlays/src/useOverlay.ts:119]: can you just add a link to the issue here?

File: packages/@react-spectrum/overlays/src/Modal.tsx [hunk 6]
```diff
@@ -26,25 +26,25 @@ interface ModalWrapperProps extends HTMLAttributes<HTMLElement> {
...
+  overlayProps: HTMLAttributes<HTMLElement>
 }
 
 function Modal(props: ModalProps, ref: DOMRef<HTMLDivElement>) {
-  let {children, onClose, type, isDismissable, isKeyboardDismissDisabled, ...otherProps} = props;
+  let {children, onClose, type, ...otherProps} = props;
   let domRef = useDOMRef(ref);
   let {styleProps} = useStyleProps(props);
 
+  let {overlayProps, underlayProps} = useOverlay(props, domRef);
```
Comment (devongovett) [inline packages/@react-spectrum/overlays/src/Modal.tsx:37]: Probably also need this change in Tray?

File: packages/@react-aria/overlays/src/useOverlay.ts [hunk 7]
```diff
@@ -46,7 +46,9 @@ interface OverlayProps {
 
 interface OverlayAria {
   /** Props to apply to the overlay container element. */
-  overlayProps: HTMLAttributes<HTMLElement>
+  overlayProps: HTMLAttributes<HTMLElement>,
+  /** Props to apply to the underlay element. */
```
Comment (devongovett) [inline packages/@react-aria/overlays/src/useOverlay.ts:50]:
```suggestion
  /** Props to apply to the underlay element, if any. */
```

File: packages/@react-spectrum/overlays/src/Modal.tsx (hunk 6 above)
Comment (snowystinger) [reply packages/@react-spectrum/overlays/src/Modal.tsx:37]: good catch

**Review by devongovett — APPROVED** [review]

**Pull request merged**